File: web/js/motion.js
/**
 * Per-key motion detection - a port of src/slicedeck/motion.py.
 *
 * Two signals, because either alone misses a class of motion:
 *   mean difference  - catches whole-tile changes, blind to small fast movers
 *   changed area     - catches the small mover, blind to uniform drift
 * A key is moving if either fires; a key is dirty (worth redrawing) at much
 * lower bars. Skipping clean keys is what keeps the frame budget flat as the
 * grid grows from 6 keys to 32.
 */

export const PROBE = 24;

export class MotionDetector {
  constructor(options = {}) {
    this.threshold = options.threshold ?? 12;
    this.pixelFloor = options.pixelFloor ?? 10;
    this.motionArea = options.motionArea ?? 0.02;
    this.dirtyThreshold = options.dirtyThreshold ?? 1;
    this.dirtyArea = options.dirtyArea ?? 0.004;

    this.probes = new Map();
    // One reusable scratch canvas: allocating per key per frame is what makes
    // naive canvas pipelines stutter once the grid gets large.
    this.scratch = document.createElement('canvas');
    this.scratch.width = PROBE;
    this.scratch.height = PROBE;
    this.ctx = this.scratch.getContext('2d', { willReadFrequently: true });
  }

  reset() { this.probes.clear(); }

  /** Downsample a tile to a PROBE x PROBE luminance array. */
  probe(source, sx, sy, sw, sh) {
    this.ctx.drawImage(source, sx, sy, sw, sh, 0, 0, PROBE, PROBE);
    const { data } = this.ctx.getImageData(0, 0, PROBE, PROBE);
    const luma = new Uint8Array(PROBE * PROBE);
    for (let i = 0, p = 0; i < data.length; i += 4, p++) {
      // Rec. 601 luma, matching PIL's ImageOps.grayscale.
      luma[p] = (data[i] * 299 + data[i + 1] * 587 + data[i + 2] * 114) / 1000;
    }
    return luma;
  }

  update(key, source, sx, sy, sw, sh) {
    const current = this.probe(source, sx, sy, sw, sh);
    const previous = this.probes.get(key);
    this.probes.set(key, current);

    if (!previous) {
      return { score: 255, area: 1, moving: true, dirty: true };
    }

    let total = 0;
    let changed = 0;
    for (let i = 0; i < current.length; i++) {
      const diff = Math.abs(current[i] - previous[i]);
      total += diff;
      if (diff > this.pixelFloor) changed++;
    }
    const score = total / current.length;
    const area = changed / current.length;

    return {
      score,
      area,
      moving: score >= this.threshold || area >= this.motionArea,
      dirty: score >= this.dirtyThreshold || area >= this.dirtyArea,
    };
  }
}

File: src/slicedeck/motion.py
"""Per-key motion detection and change tracking.

Two jobs, one cheap frame difference:

* **Motion** - which keys changed enough to be worth flagging in the UI.
* **Dirty tracking** - which keys changed at all. Only those get re-encoded and
  rewritten, which is where most of the pipeline's savings come from: a static
  scene costs one JPEG encode instead of ``cols * rows`` of them.

Scoring uses two signals, because either one alone gets a class of motion wrong:

* **mean difference** catches whole-tile changes (lights switching on, exposure
  shifts) but a small fast object averages away to nothing across a big tile;
* **changed area** - the share of pixels that moved by more than the noise floor
  - catches the small object but ignores a uniform brightness drift.

A key is moving if *either* fires.
"""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np
from PIL import Image, ImageOps

#: Tiles are downsampled to this many pixels per side before comparison, so
#: sensor noise and JPEG artefacts average out and the check stays cheap.
PROBE = 24


@dataclass
class KeyState:
    #: Mean absolute luminance difference from the previous frame, 0-255.
    score: float = 0.0
    #: Share of the tile whose pixels moved past the noise floor, 0-1.
    area: float = 0.0
    moving: bool = False
    dirty: bool = True


@dataclass
class MotionDetector:
    """Tracks each key's previous tile and scores the new one against it."""

    #: Mean difference that counts as motion on its own.
    threshold: float = 12.0
    #: Per-pixel difference below this is treated as noise, not movement.
    pixel_floor: float = 10.0
    #: Changed area that counts as motion on its own.
    motion_area: float = 0.02
    #: Below these, a tile is considered unchanged and is not re-encoded.
    dirty_threshold: float = 1.0
    dirty_area: float = 0.004

    _probes: dict[int, np.ndarray] = field(default_factory=dict, repr=False)

    def reset(self) -> None:
        """Forget every key, so the next frame redraws all of them."""
        self._probes.clear()

    @staticmethod
    def _probe(tile: Image.Image) -> np.ndarray:
        small = ImageOps.grayscale(tile).resize((PROBE, PROBE), Image.Resampling.BILINEAR)
        return np.asarray(small, dtype=np.int16)

    def update(self, key: int, tile: Image.Image) -> KeyState:
        probe = self._probe(tile)
        previous = self._probes.get(key)
        self._probes[key] = probe

        if previous is None:
            # First sighting of this key: nothing to compare against, so draw it.
            return KeyState(score=255.0, area=1.0, moving=True, dirty=True)

        diff = np.abs(probe - previous)
        score = float(diff.mean())
        area = float((diff > self.pixel_floor).mean())

        return KeyState(
            score=score,
            area=area,
            moving=score >= self.threshold or area >= self.motion_area,
            dirty=score >= self.dirty_threshold or area >= self.dirty_area,
        )
